edit-post/components/sidebar/featured-image/index.js

```diff
@@ -44,4 +51,17 @@ export default connect(
 	},
 	undefined,
 	{ storeKey: 'edit-post' }
+);
+
+const applyWithAPIData = withAPIData( ( props ) => {
+	const { postTypeSlug } = props;
+	return {
+		postType: `/wp/v2/types/${ postTypeSlug }?context=edit`,
+	};
+} );
```

Comment:
We should definitely settle all together on a unique way to do this. Whether we use apply* functions or just embed these function in `compose`. I know @mcsf had some thoughts on this.

Reply:
(What follows is just a stream of thought.) Some of the concerns with `apply*` are verbosity but especially the risk of creating something too opaque for a newcomer. Without the long framework experience some of us have had the chance to accrue, what does `compose( applyQuery, applyConnect, applyWithAPIData )` mean?

This doesn't mean that inlining these HoCs is necessarily better; I think they could bring some clarity when they are small enough. Ideally, the more similar their interfaces are—aside from the _source_ they refer to—the better, reducing noise and cognitive burden.

Could there be different approaches? For instance, is it conceivable that, in most cases, partially applied HoCs can be named based on the data they seek to retrieve rather than the sourcing mechanism? For instance (and don't pay too much attention to the `with` prefix):

```js
compose(
  withPostType,
  withSidebarState,
)
```

Above, I tentatively folded two different operations under `withPostType`:

```js
withPostType = compose(
  query( ( select ) => ( { postTypeSlug: select( … ) } ),
  withAPIData( ( { postTypeSlug } ) => ( { postType: `/wp…` } ) ),
)
```

* * *

Naturally, an issue here becomes performance. If my data needs concern different domains, we could end up with:

```js
compose(
  withPostType, // composes query + withAPIData
  withImageOfTheDay, // composes connect + withAPIData
)
```

… where we end up with very deep and redundant wrapping of components:

```
query -> withAPIData -> connect -> withAPIData -> WrappedComponent
```

If that were the case, I think we'd able to converge on the right providers and composing helpers to make this smooth enough—usage-wise and performance-wise.